File: ProdutosAngular/src/app/model/produto.ts
import { Fornecedor } from "./fornecedor";


export class Produto {
  public id: number;

  public idUsuario: number;

  public nome: string;

  public descricao: string;

  public frete: number;

  public promocao: boolean;

  public valorTotalDesc: number;

  public valorTotalFrete: number;

  public valor: number;

  public valorInicial: number;

  public quantia: number;

  public somaTotalValores: number;

  public freteAtivo: boolean;

  public valorDesconto: number;

  public imagem: string;

  public fornecedor: Fornecedor | null;

  constructor(
    id: number,
    idUsuario: number,
    nome: string,
    descricao: string,
    frete: number,
    promocao: boolean,
    valorTotalDesc: number,
    valorTotalFrete: number,
    valor: number,
    valorInicial: number,
    quantia: number,
    somaTotalValores: number,
    freteAtivo: boolean,
    valorDesconto: number,
    imagem: string,
    fornecedor: Fornecedor | null
  ) {
    this.valorDesconto = valorDesconto;
    this.idUsuario = idUsuario;
    this.freteAtivo = freteAtivo;
    this.somaTotalValores = somaTotalValores;
    this.quantia = quantia;
    this.valor = valor;
    this.valorInicial = valorInicial;
    this.valorTotalFrete = valorTotalFrete;
    this.valorTotalDesc = valorTotalDesc;
    this.promocao = promocao;
    this.frete = frete;
    this.descricao = descricao;
    this.nome = nome;
    this.id = id;
    this.imagem = imagem;
    this.fornecedor = fornecedor;
  }
}

File: produto-service/src/main/java/com/prodify/produto_service/domain/model/Produto.java
package com.prodify.produto_service.domain.model;

import java.math.BigDecimal;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Lob;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.NamedAttributeNode;
import jakarta.persistence.NamedEntityGraph;
import jakarta.persistence.NamedSubgraph;
import jakarta.persistence.SequenceGenerator;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.NotNull;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import com.prodify.produto_service.domain.model.Fornecedor;

@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
@ToString
@SequenceGenerator(name = "produto_seq", sequenceName = "produto_sequence", allocationSize = 1)
@Schema(name = "Produto Entity")
@Builder
public class Produto {

	@Id
	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "produto_seq")
	@Schema(description = "ID do produto - vem null do frontend, gerado automaticamente pelo JPA para novos produtos, obrigatório para atualizações")
	private Long id;

	@Column(nullable = false)
	@NotNull(message = "O ID Usuário é obrigatório")
	@Schema(description = "ID do usuário")
	private Long idUsuario;

	@Column(length = 100, nullable = false)
	@NotNull(message = "O nome é obrigatório")
	@Schema(description = "Nome do produto")
	private String nome;
	
	@Lob
	@Column(name = "imagem", nullable = true)
	@Schema(description = "Imagem do produto")
	private byte[] imagem;
	
	@Column(length = 5, nullable = false)
	@NotNull(message = "Promoção é obrigatório")
	@Schema(description = "Se o produto está em promoção")
	private Boolean promocao;
	
	@Column(nullable = true, length = 100)
	@Schema(description = "Valor total com desconto")
	private BigDecimal valorTotalDesc;

	@Column(nullable = true, length = 100)
	@Schema(description = "Soma total dos valores")
	private BigDecimal somaTotalValores;
	
	// Relação Muitos-para-Um
    @ManyToOne
    @JoinColumn(name = "fornecedor_id", nullable = false)
	@NotNull(message = "Fornecedor é obrigatório")
    @Schema(description = "Fornecedor do produto")
    private Fornecedor fornecedor;
	
	@Column(nullable = true, length = 100)
	@Schema(description = "Valor do produto")
	private BigDecimal valor;

	@Column(nullable = true, length = 100)
	@Schema(description = "Valor total do frete")
	private BigDecimal valorTotalFrete;

	@Column(nullable = true, length = 100)
	@NotNull(message = "Descrição do produto é obrigatória")
	@Schema(description = "Descrição do produto")
	private String descricao; 
	
	@Column(nullable = true, length = 100)
	@NotNull(message = "Valor do frete é obrigatório")
	@Schema(description = "Valor do frete")
	private BigDecimal frete; 
	
	@Column(nullable = true, length = 100)
	@Schema(description = "Valor inicial")
	@Max(value = 2000000000, message = "O valor inicial não pode ser maior que 2.000.000.000")
	private BigDecimal valorInicial;

	@Column(nullable = true, length = 100)
	@Schema(description = "Quantidade")
	@Max(value = 1000000, message = "A quantidade não pode ser maior que 1.000.000")
	private Long quantia;

	@Column(nullable = false, length = 5)
	@NotNull(message = "Frete ativo é obrigatório")
	@Schema(description = "Se o frete está ativo")
	private Boolean freteAtivo; 

	@Schema(description = "Valor do desconto")
	@Column(nullable = true, length = 100)
	private BigDecimal valorDesconto; 
}